[Blog] Improve OpenSearch cluster performance by separating search and indexing workloads (#3772)
opensearch-project/project-website: _posts/2025-05-09-improve-opensearch-cluster-performance.md

---
layout: post
title: "Improve OpenSearch cluster performance by separating search and indexing workloads"
authors:
  - naarcha

>   - pudyoda

Author: ```suggestion
  - pudyodu
```

> date: 2025-05-09

Reviewer: Update to: 2025-05-27

categories:
  - technical-posts

> meta_keywords: OpenSearch, shard roles, indexing, search separation, performance tuning

Author: meta_keywords: OpenSearch cluster performance, index shards, search workloads, node roles, indexing and search separation, OpenSearch benchmark, performance tuning, optimizing your cluster

> meta_description: Learn how separating indexing and search workloads in OpenSearch can improve performance, scalability, and cost efficiency using dedicated roles and node types.

Reviewer: Update meta with the following: 

_meta_keywords:_ OpenSearch cluster performance, index shards, search workloads, node roles, indexing and search separation, OpenSearch benchmark, performance tuning, optimizing your cluster

_meta_description:_  Optimize OpenSearch cluster performance by separating indexing and search operations. Discover write/search replicas and dedicated search nodes, with benchmark data showing 34% lower query latency through workload separation.

> meta_description: Learn how separating indexing and search workloads in OpenSearch can improve performance, scalability, and cost efficiency using dedicated roles and node types.

Author: ```suggestion
meta_description: Optimize OpenSearch cluster performance by separating indexing and search operations. Discover write/search replicas and dedicated search nodes, with benchmark data showing 34% lower query latency through workload separation.
```

- The write replica downloads segments after the primary completes writing.
- The search replicas continuously poll the remote store for new segments.

![OpenSearch cluster architecture for search and indexing separation](/assets/media/blog-images/2025-05-09-search-index-separation/rw-separation-architecture.png)

## Benefits 

Separating indexing and search workloads provides the following benefits:

* **Parallel and isolated processing**: Improve throughput and predictability by isolating indexing and search.
* **Independent scalability**: Scale indexing and search independently by adding `data` or `search` nodes.
* **Failure resilience**:  Indexing and search failures are isolated, improving availability.
* **Cost efficiency and performance**: Use specialized hardware—compute-optimized for indexing, memory-optimized for search.
* **Tuning flexibility**: Optimize performance settings like buffers and caches independently for indexing and search.
* **Failure isolation**: Indexing and search issues are separated, which helps troubleshoot failures
* **Scalable design**: Scale indexing and search independently of each other

## Enabling indexing and search separation

> For detailed instructions, see [Separate index and search workloads](https://docs.opensearch.org/docs/latest/search-plugins/index-management/index-and-search-separation/).

Reviewer 2: ```suggestion
For detailed instructions, see [Separate index and search workloads](https://docs.opensearch.org/docs/latest/tuning-your-cluster/seperate-index-and-search-workloads/).
```

## Scale to zero with reader/writer separation: The search-only mode

In write-once, read-many scenarios—like log analytics or frozen time-series data—you can reduce resource use after indexing completes. OpenSearch now supports a **search-only mode** through the `_scale` API, allowing you to disable primary shards and write replicas, leaving only search replicas active. 

This significantly reduces storage and compute costs while maintaining full search capabilities and provides the following benefits:

- Scale down indexing capacity when writes are no longer needed
- Free up disk and memory by removing unnecessary write paths
- Work smoothly with index lifecycle controls through `_open`/`_close` operations
- Keep only search replicas active to reduce resource usage
- Maintain green cluster health with active search paths
- Auto-recover search replicas without manual intervention
- Scale back up at any time to resume indexing
- Scale search replicas based on search traffic during search-only mode

> When `search-only` mode is active, OpenSearch maintains cluster health as GREEN when search-only mode is active despite disabling primaries and write replicas, This is because only search replicas are expected to be allocated, and they are all successfully assigned. This ensures operational stability and full search availability.

Reviewer 2: ```suggestion
OpenSearch maintains cluster health as GREEN when search-only mode is active despite disabling primaries and write replicas. This is because only search replicas are expected to be allocated, and they are all successfully assigned. This ensures operational stability and full search availability.
```

### How it works

> When the `_scale` API is called with `{ "search_only": true }`, OpenSearch performs the following index:

Reviewer 2: ```suggestion
When the `_scale` API is called with `{ "search_only": true }`, OpenSearch performs the following operations:
```

1. Adds an internal search-only block to the index
2. Scales down all primaries and write replicas
3. Keeps only search replicas active

> To resume indexing, run the `_scale` operation with `{ "search_only": false }`. OpenSearch restores the original index state. Even in search-only mode, cluster health remains green because all expected search replicas are allocated.

Reviewer 2: ```suggestion
To resume indexing, run the `_scale` operation with `{ "search_only": false }`. OpenSearch restores the original index state. Even in search-only mode, cluster health remains GREEN because all expected search replicas are allocated.
```

## Benchmark comparison

We benchmarked indexing throughput and query latency across three cluster configurations using the `http_logs` workload in OpenSearch Benchmark. The test ran 50% of indexing first, followed by simultaneous indexing and an expensive multi-term aggregation query.

### Cluster configurations

> Using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk), we set up three different clusters in AWS. The followings table show the node roles and associated costs based on the EC2 instance types used. Instance pricing is from [AWS EC2 instance types](https://aws.amazon.com/ec2/instance-types/).

Reviewer 2: ```suggestion
Using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk), we set up three different clusters on AWS. The followings table shows the node roles and associated costs based on the Amazon Elastic Compute Cloud (Amazon EC2) instance types used. Instance pricing is taken from [Amazon EC2 Instance types](https://aws.amazon.com/ec2/instance-types/).
```

> #### Cluster 1 (c1-r6g): Standard Cluster with Memory optimized instances for data nodes

Reviewer 2: ```suggestion
#### Cluster 1 (c1-r6g): Standard cluster with Memory Optimized instances for data nodes
```

> | Node Role	| Number of Nodes	| Instance Type	| Hourly Cost	| Total Hourly Cost	|

Reviewer 2: ```suggestion
| Node role	| Number of nodes	| Instance type	| Hourly cost	| Total hourly cost	|
```

|---	|---	|---	|---	|---	|
|Data	|4	|r6g.xlarge	|$0.20	|$0.81	|
|Coordinator	|2	|r6g.xlarge	|$0.20	|$0.40	|

> |Cluster Manager	|3	|c6g.xlarge	|$0.14	|$0.41	|

Reviewer 2: ```suggestion
|Cluster manager	|3	|c6g.xlarge	|$0.14	|$0.41	|
```

|	|	|	|	|$1.62	|

> #### Cluster 2 (c2-c6g): Standard Cluster with Compute optimized instances for data nodes 

Reviewer 2: ```suggestion
#### Cluster 2 (c2-c6g): Standard cluster with Compute Optimized instances for data nodes 
```

> | Node Role	| Number of Nodes	| Instance Type	| Hourly Cost	| Total Hourly Cost	|

Reviewer 2: ```suggestion
| Node role	| Number of nodes	| Instance type	| Hourly cost	| Total hourly cost	|
```

|---	|---	|---	|---	|---	|
|Data	|4	|c6g.xlarge	|$0.14	|$0.54	|
|Coordinator	|2	|r6g.xlarge	|$0.20	|$0.40	|

> |Cluster Manager	|3	|c6g.xlarge	|$0.14	|$0.41	|

Reviewer 2: ```suggestion
|Cluster manager	|3	|c6g.xlarge	|$0.14	|$0.41	|
```

|	|	|	|	|$1.36	|

> #### Cluster 3 (c3-indexing-search):  Indexing and Search separated cluster with Compute and Memory optimized instances for data and search nodes

Reviewer 2: ```suggestion
#### Cluster 3 (c3-indexing-search): Indexing-and-search-separated cluster with Compute and Memory Optimized instances for data and search nodes
```

| Node Role	| Nubmer of Nodes	| Instance Type	| Hourly Cost	| Total Hourly Cost	|
|---	|---	|---	|---	|---	|
|Data	|2	|c6g.xlarge	|$0.14	|$0.27	|
|Search	|2	|r6g.xlarge	|$0.20	|$0.40	|
|Coordinator	|2	|r6g.xlarge	|$0.20	|$0.40	|

> |Cluster Manager	|3	|c6g.xlarge	|$0.14	|$0.41	|

Reviewer 2: ```suggestion
|Cluster manager	|3	|c6g.xlarge	|$0.14	|$0.41	|
```

|	|	|	|	|$1.49	|

### Result visualizations

> The following graphs shows the indexing throughput and the query latency differences in these three clusters. 

Reviewer 2: ```suggestion
The following charts show the indexing throughput and the query latency differences across these three clusters. 
```

> The following chart compares indexing throughput across the three cluster configurations:

Reviewer 2: ```suggestion
The following chart compares indexing throughput across the three cluster configurations.
```

![Indexing throughput comparison](/assets/media/blog-images/2025-05-09-search-index-separation/indexing-throughput-compare-run.png)

> The chart following shows how separating workloads improves query latency:

Reviewer 2: ```suggestion
The following chart shows how separating workloads improves query latency.
```

![Query latency comparison](/assets/media/blog-images/2025-05-09-search-index-separation/query-latency.png)

> The results show that Cluster 3 slightly outperforms others for simultaneous indexing and search, while significantly reducing query latency and lowering cost.

Reviewer 2: ```suggestion
The results show that Cluster 3 slightly outperforms the other clusters in simultaneous indexing and search while significantly reducing query latency and lowering costs.
```

### Result metrics and comparisons

> The following tables provided comparisons between each cluster's benchmark results:

Reviewer 2: ```suggestion
The following tables provide comparisons between each cluster's benchmark results.
```

**Comparing Cluster 1 and Cluster 3**

|	|	|Cluster 1(5P, 3R)	|Cluster 3 (5P, 1R, 1S)	|Difference	|
|---	|---	|---	|---	|---	|

> |Index Throughput	|Median	|58523.95095	|59818.80119	|2.16462	|

Reviewer 2: ```suggestion
|Index throughput	|Median	|58523.95095	|59818.80119	|2.16462	|
```

|	|Maximum	|58777.10938	|60240.02344	|2.42848	|
|	|	|	|	|	|

> |Query Latency	|p50	|6725.72925	|5358.19263	|-25.52235	|

Reviewer 2: ```suggestion
|Query latency	|p50	|6725.72925	|5358.19263	|-25.52235	|
```
